### ISSUE TITLE
ideas

### ISSUE THREAD

**Issue by hrsh7th** [hrsh7th/nvim-compe#120]
Feel free to post ideas. (I don't know if i do)

**Comment by hrsh7th** [comment]
git source.

**Comment by tami5** [comment]
> git source.

hmm interesting

**Comment by hrsh7th** [comment]
I don't know that is this source if teally useful.
But maybe all branch name or log (with docemtation)

**Comment by hrsh7th** [comment]
project files source. (provodes all relative file paths)

**Comment by hrsh7th** [comment]
register or yank history source

**Comment by ii14** [comment]
> project files source. (provodes all relative file paths)

Maybe there should be an option for `path` source to include files from the current git repo?

**Comment by ii14** [comment]
The [necosyntax](https://github.com/tamago324/compe-necosyntax) is really cool, but it has a dependency and I noticed that it slows down vim startup time. Maybe rewriting it to lua and somehow doing it asynchronously would help? I'll try to look into that when I have some free time.

**Comment by cloggier** [comment]
> register or yank history source

Getting yanks from miniyank perhaps?

**Comment by cloggier** [comment]
How about lbdb source to complete email addresses from address book while composing emails?

**Comment by ii14** [comment]
> The necosyntax is really cool, but it has a dependency and I noticed that it slows down vim startup time. Maybe rewriting it to lua and somehow doing it asynchronously would help? I'll try to look into that when I have some free time.

Nope, I got it wrong, just checked it again and it works just fine. Sorry.

**Comment by ii14** [comment]
@hrsh7th I know that you've said you don't like completions from other buffers, but some people like to have that so maybe it makes sense to provide that?

**Comment by hrsh7th** [comment]
I don't want to provide settings as much as possible.

In my random thoughts, we provide specific prefix 
for triggering some source.

For example,
1. !/ -> project files
2. @. -> yank-history

But I don't know if these useful 😅

**Comment by hrsh7th** [comment]
Probably, we shoyld design source option interface.

**Comment by ii14** [comment]
I don't know if it would be actually useful for anyone, but what about `compe#completion()` optionally taking in source as argument?
You could for example create mappings like builtin `<C-X><C-F>`:
```vim
inoremap <silent><expr> <C-X><C-F> compe#complete('path')
```

**Comment by hrsh7th** [comment]
It' good idea.
I will consider it.

**Comment by tamago324** [comment]
> git source.

What is git source, and is it something like https://github.com/zk-phi/git-complete?
I don't use emacs, so I don't know what it's like, but it would be great if it could!

**Comment by tami5** [comment]
> > git source.
> 
> What is git source, and is it something like https://github.com/zk-phi/git-complete?
> I don't use emacs, so I don't know what it's like, but it would be great if it could!

oh wow :heart_eyes:

**Comment by ii14** [comment]
> > git source.
> 
> What is git source, and is it something like https://github.com/zk-phi/git-complete?
> I don't use emacs, so I don't know what it's like, but it would be great if it could!

This is really cool and deserves a vim port in its own right, but is nvim-compe a right plugin for this? From the README it seems that you have to first write a "query" and only then you call the completion to do the grep through your project and find common patterns. This could be as well implemented as a command like `:GitComplete {pattern}`. Auto-completion plugins like nvim-compe on the other hand suggest you completions on the fly, as you type.

**Comment by cloggier** [comment]
- Treesitter source
- Configuration for sources depending on cursor position. Like completion-nvim which allows to, for example, limit sources for comments. 

Not sure if that would come with performance hit.

**Comment by ii14** [comment]
> Configuration for sources depending on cursor position. Like completion-nvim which allows to, for example, limit sources for comments.

I didn't know that completion-nvim can do that, but it probably can be solved by doing a regex match on the current line.

**Comment by elianiva** [comment]
> * Treesitter source
> * Configuration for sources depending on cursor position. Like completion-nvim which allows to, for example, limit sources for comments.
> 
> Not sure if that would come with performance hit.

I've made treesitter source but it's a bit clunky, I'll try to come back to that and see if I can improve it.

**Comment by bluz71** [comment]
[vim-lsc](https://github.com/natebosch/vim-lsc) source, similar to the existing [vim-lsp](https://github.com/prabirshrestha/vim-lsp) source.

Why? Because Neovim's current implementation of [`allow_incremental_sync`](https://github.com/neovim/neovim/pull/13371) does not work for dot-completion via the [Solargraph](https://github.com/castwide/solargraph) Ruby language server. [As explained here](https://github.com/neovim/neovim/issues/13049#issuecomment-772217349) Neovim LSP is sending full lines for `didChange` whilst vim-lsc is sending minimal insert text `didChange`. Basically vim-lsc has working Ruby dot-completions and Neovim LSP currently does not.

So I wish to use vim-lsc as a source for nvim-compe but just for the Ruby language whilst using Neovim LSP for JavaScript and Dart, for example, something like:

```lua
local compe = require'compe'

compe.setup {
  enabled = true,
  min_length = 2,
  preselect = 'disable',

  source = {
    buffer = {
      filetypes = {'dart', 'javascript', 'json', 'ruby'}
    },
    nvim_lsp = {
      filetypes = {'dart', 'javascript'}
    },
    vim_lsc = {
      filetypes = {'ruby'}
    },
    vsnip = {
      filetypes = {'dart', 'javascript', 'ruby'}
    }
  }
}
```

**Comment by hrsh7th** [comment]
The vim-lsc source is easy and reasonable. (but it might not be perfect if it has a capabilities problem)

I will add it.

**Comment by hrsh7th** [comment]
Other sources need more consideration for detailed design and experience.

**Comment by bluz71** [comment]
> The vim-lsc source is easy and reasonable. 

Fantastic. Many thanks, it will solve a real problem for me.

**Comment by hrsh7th** [comment]
@bluz71 I've added vim-lsc source. if you met problem, please create PR.

**Comment by bluz71** [comment]
@hrsh7th,

I have tested the vim-lsc integration and it appears to work well. I am very happy about that.

The cool thing is I can now use different LSP engines for different filetypes. This is my setup for those interested:

```lua
local compe = require'compe'

compe.setup {
  enabled = true,
  min_length = 2,
  preselect = 'disable',

  source = {
    buffer = {
      filetypes = {'eruby', 'dart', 'html', 'javascript', 'json', 'ruby'}
    },
    nvim_lsp = {
      filetypes = {'dart', 'eruby', 'html', 'javascript'}
    },
    vim_lsc = {
      filetypes = {'ruby'}
    },
    vsnip = {
      filetypes = {'dart', 'html', 'javascript', 'markdown', 'ruby'}
    }
  }
}
```

Many thanks.

**Comment by elianiva** [comment]
@cloggier you could try treesitter source on my fork. https://github.com/elianiva/nvim-compe/tree/feat/treesitter_source

I ported `nvim-treesitter/completion-treesitter` so the implementation is pretty much the same with some little changes here and there. I can't fully confirm if the behaviour is the same though since I never tried treesitter source for completion.nvim

**Comment by lucax88x** [comment]
I'm thinking about any possibily to "limit" or "switch" sources.

Example:

Let's say I have these sources:

- nvim-lsp
- vim-snip

I write "for".

lsp gives me TONS of suggestions for sentence 'for' because yes (for, fork, format, etc... almost 50 in my case) and only one for snippets (because I have a "for" snippet)

I need to scroll ALL the lsp suggestions before eventually and finally finding my snippets.

Maybe could be a good idea to "limit suggestions" from each source? Or maybe possibility to switch between sources with a shortcut?

**Comment by hrsh7th** [comment]
Good concern.
Agree with you but I want `ideal design` for it.

For example, I have many patterns to solve it.

1. Add mapping to `confirm`

`inoremap <expr><CR> compe#confirm()`
`inoremap <expr><S-CR> compe#confirm('vsnip')`

2. Add mapping to `complete`

`inoremap <expr><C-Space> compe#complete()`
`inoremap <expr><C-s> compe#complete('vsnip')`

3. Add mapping to `<C-n>`
`inoremap <expr><C-j> compe#next_source()`

**Comment by lucax88x** [comment]
1. What would this do? Confirm with first snip found?
If I have 50 items from a LSP source, I don't even see the vsnip suggestion so I don't know what I am confirming

2. I don't like this one, if I have 4 sources I can't have 4 mappings for that, could be a bit cumbersome.

3. IMHO, this is the way, also dividing in "groups" could lead to performance benefits (querying one source at a time)

**Comment by cloggier** [comment]
One option would be to implement chain completion, like completion.nvim and [lifepillar/vim-mucomplete](https://github.com/lifepillar/vim-mucomplete). I was never big fan of those so would really dislike it becoming the only option for compe, but as an configurable option it could be a solution to @lucax88x's problem.

**Comment by elianiva** [comment]
> 1. What would this do? Confirm with first snip found?
>    If I have 50 items from a LSP source, I don't even see the vsnip suggestion so I don't know what I am confirming
> 2. I don't like this one, if I have 4 sources I can't have 4 mappings for that, could be a bit cumbersome.
> 3. IMHO, this is the way, also dividing in "groups" could lead to performance benefits (querying one source at a time)

I agree with the third option, I don't always want all sources to be there so it would be nice to be able to toggle between them.

**Comment by hrsh7th** [comment]
@cloggier I have the same opinion. I don't like chain completion.

I will create PR to show `next_source` mapping as a proof of concepts.

**Comment by elianiva** [comment]
by the way, @cloggier are you satisfied with the treesitter source? If we *do* want nvim-compe to support treesitter source and you like that implementation, I'll submit a PR.

**Comment by cloggier** [comment]
@elianiva Frankly I never used TS completion before so not sure if I can reliably test. Maybe just submit a PR and someone with more TS knowledge can pick it up?

**Comment by hrsh7th** [comment]
I have another idea to choose a specific source item.

If we provide `sticky_char` option to source's metadata and set `vsnip = { sticky_char = 0 }`, we can choose vsnip item `for0`

**Comment by elianiva** [comment]
that's also good IMO, so we don't have to cycle every single one of the available sources

**Comment by lucax88x** [comment]
If by sticky you mean "confirming the next source item" I still see the same issue I mentioned before

```quote
If I have 50 items from a LSP source, I don't even see the vsnip suggestion so I don't know what I am confirming
```

**Comment by hrsh7th** [comment]
No. `sticky_char` can use filter. If you type `for0` with the above settings, compe will suggest vsnip source (maybe) only.

**Comment by lucax88x** [comment]
Ok, Yeah, I like that! 

Could we possibly have problems that maybe someone really wants to write for0?

maybe better... use C-0?

**Comment by hrsh7th** [comment]
If the buffer has `item0`, `item2`, `item3` as contents, this solution may not work as expected.
But I don't think there is such a problem in real-world because the completion-menu will be enough small if we type `0`.

Anyway, it is easy to implement. I try to use it (and will sent PR)

**Comment by cloggier** [comment]
I am not sure I like it. Perhaps some function to filter results which then could be mapped to a key by user?
Something like `v:lua.compe.filter{only = {'vsnip'}}` for example?

**Comment by hrsh7th** [comment]
`<C-0>` needs extra mapping so it will be complex and might buggy.

**Comment by hrsh7th** [comment]
The `sticky_char` idea is only modify the item's `word` like `%WORD%_%STICKY_CHAR%`.
In compe, we can filter `item1` by `i1` so it is the same as this.

**Comment by hrsh7th** [comment]
It's easy...

Uploading Kapture 2021-02-04 at 21.54.15.mp4…

**Comment by lucax88x** [comment]
I can try your branch?

**Comment by hrsh7th** [comment]
I've sent PR.
Please feedback it.

**Comment by tzachar** [comment]
Hi.

Would be extra helpfull to have a TabNine source. 
I don't think my Lua-foo is good enough to implement it though.... 
A good example of a python implementation is tbodt/deoplete-tabnine

On a side note, is there any documentation on the execution model of lua extentions inside neovim? and inside nvim-compe?

**Comment by hrsh7th** [comment]
### sources
- git-grep
- project-files
- yank-history
- lbdb?
- ~~treesitter~~ #133 
- tabnine

### features
- source params interface
- invoke specific source completion
- source configurations(syntax/ignored_syntaxes)
- to select specific item easily

**Comment by tami5** [comment]
> ### sources
> 
> - git-grep
> 
> - project-files
> 
> - yank-history
> 
> - lbdb?
> 
> - ~~treesitter~~ #133 
> 
> - tabnine
> 
> 
> 
> ### features
> 
> - source params interface
> 
> - invoke specific source completion
> 
> - source configurations(syntax/ignored_syntaxes)
> 
> - to select specific item easily
> 

@hrsh7th 😍😍😍😍 You are the best


commandline completion, is it possible to todo so ?

**Comment by hrsh7th** [comment]
I don't know how to customize cmdline completion in Lua/Vim script.

You can use cmdline-window (default mapping is `q:`)

**Comment by tami5** [comment]
Oh great suggestion

**Comment by cloggier** [comment]
Also fzf.vim comes with `:Commands`.

**Comment by ii14** [comment]
`:help getcompletion()`?

**Comment by elianiva** [comment]
> commandline completion, is it possible to todo so ?

Command line completion as in, compe in cmdline or cmdline source? I think what you meant is the latter but I don't want to make assumption :laughing:

If it is, then something simple like `getcompletion('he', 'command')` and feed the result to nvim-compe should be enough.

**Comment by ii14** [comment]
> commandline completion, is it possible to todo so ?

> Also fzf.vim comes with `:Commands`.

> If it is, then something simple like `getcompletion('he', 'command')` and feed the result to nvim-compe should be enough.

Maybe just a vimscript source based on `getcompletion()`? But it would make more sense to do it in a separate repo for now.

**Comment by ii14** [comment]
It would be really cool if we could create a language server based on `nvim --headless` by somehow sandboxing it or revoking any write permissions so it won't screw up anything on your system and just letting it evaluate everything. (Neo)vim provides some functionality that kinda allows to do that, like the already mentioned `getcompletion()`, but also `nvim_parse_expression()`, `:verbose`, etc.

**Comment by bresilla** [comment]
What about adding icons to each completion entry, just like: https://github.com/onsails/lspkind-nvim does...

**Comment by cloggier** [comment]
@bresilla You can set it up yourself.

```lua
local protocol = require("vim.lsp.protocol")

protocol.CompletionItemKind = {…}
```

Good place for is your custom `on_attach` function if you have one.

**Comment by bresilla** [comment]
> @bresilla You can set it up yourself.
> 
> ```lua
> local protocol = require("vim.lsp.protocol")
> 
> protocol.CompletionItemKind = {…}
> ```
> 
> Good place for is your custom `on_attach` function if you have one.

Thanks @cloggier , i am already using it, but that's just for LSP sources. I was thining more in lines with adding icons to each source, like TabNine, or other sources that nvim-compe can serve...

Correct me if i am wrong, but you can't do that with `require("vim.lsp.protocol")`??

**Comment by cloggier** [comment]
You can configure each source and set what gets displayed as label by using `menu` key. Is that what you are after?

**Comment by bresilla** [comment]
> You can configure each source and set what gets displayed as label by using `menu` key. Is that what you are after?

Yesss, that's exactly what i am after. And i don't know how to do that actually. I don't know how to use 'menu'. Would you be so kind and just tell me one example e.g. how to put a symbol in front of TabNine [TN] sources, then i do the rest. Thanks in advance.

**Comment by cloggier** [comment]
@bresilla sure, from memory: where you specify your sources and have `tabnine = true` change it to something like `tabnine = {menu = 'my custom label'}`.

**Comment by hrsh7th** [comment]
You can use `{ kind = 'you want' }` like the above example.

**Comment by echasnovski** [comment]
What I am currently missing from other completion plugins:

- Showing function signature when typing arguments. For example: type `print`, accept suggestion from LSP, type `(`. It is super convenient to have a popup window showing possible arguments of `print`. Even better If currently typed argument gets some sort of highlighting in this popup.
- "Opened buffers" source. Having current buffer source is great, but having completion from all opened buffers is even better.

**Comment by hrsh7th** [comment]
signature help is out of scope. it should be implemented as separate plugin by someone.

buffers source is discussing in #147

**Comment by echasnovski** [comment]
Thank you for your response.
About signature help: I thought that utilizing `vim.lsp.buf.signature_help()` was reasonable and feasible in compe. If it is out of scope, then no further questions :)

**Comment by matu3ba** [comment]
> > ### sources
> > 
> > * git-grep
> > * project-files
> > * yank-history
> > * lbdb?
> > * ~treesitter~ #133
> > * tabnine
> > 
> > ### features
> > 
> > * source params interface
> > * invoke specific source completion
> > * source configurations(syntax/ignored_syntaxes)
> > * to select specific item easily
> 
> @hrsh7th heart_eyesheart_eyesheart_eyesheart_eyes You are the best
> 
> commandline completion, is it possible to todo so ?

If you are interested in writing a frontend-agnostic command completion server, check out [shellac](https://gitlab.redox-os.org/AdminXVII/shellac-server/-/tree/next). Command arguments are often more complex than simple regex. This would require defining (relative simple) grammars for every CLI tool.

**Comment by matu3ba** [comment]
> I'm thinking about any possibily to "limit" or "switch" sources.
> 
> Example:
> 
> Let's say I have these sources:
> 
>     * nvim-lsp
> 
>     * vim-snip
> 
> 
> I write "for".
> 
> lsp gives me TONS of suggestions for sentence 'for' because yes (for, fork, format, etc... almost 50 in my case) and only one for snippets (because I have a "for" snippet)
> 
> I need to scroll ALL the lsp suggestions before eventually and finally finding my snippets.

This will only work with a huge numbered pop-up window or will be very annoying. The short suggestions should be listed first though. Dynamic switching will be annoying to maintain, unless it is a filter.

**Comment by tami5** [comment]
:D imagine spell source act like google gmail, suggestion the next word to use :laughing:

**Comment by hrsh7th** [comment]
https://github.com/high-moctane/nextword

**Comment by tzachar** [comment]
> https://github.com/high-moctane/nextword

A better solution would be for someone to build an open source clone of TabNine.
You can then have it learn both programming languages and prose.

**Comment by timsofteng** [comment]
@hrsh7th is it any way to launch compre in ordinary vim, not neovim?

**Comment by bluz71** [comment]
Compe is a Lua plugin that uses Neovim API.

I highly doubt compe will ever be compatible with Vim. Better that it be a great Neovim plugin that makes use of the awesome Lua JIT runtime that ships with Neovim.

Just switch over to Neovim, it is not that hard and you get access to awesome technology such as compe, Treesitter and native LSP.

**Comment by hrsh7th** [comment]
https://github.com/hrsh7th/nvim-compe/issues/1

**Comment by bluz71** [comment]
There already plenty of Vim-compatible completion managers:

- [deoplete](https://github.com/Shougo/deoplete.nvim)

- [ncm2](https://github.com/ncm2/ncm2)

- [asyncomplete.vim](https://github.com/prabirshrestha/asyncomplete.vim)

- [completor.vim](https://github.com/maralla/completor.vim)

The cool thing about nvim-compe is that hrsh7th can use all the capabilities of Neovim + API and Lua + JIT. No dumbing down for lowest common denominator.

One of the reasons that nvim-compe is as good as it is is due to the focus on Neovim 0.5. Aim for one platform and do it really well.

Well done hrsh7th and team. This is one of the very best Neovim plugins for sure.

**Comment by thallium** [comment]
Auto select the first completion entry?

**Comment by elianiva** [comment]
@thallium that's what `preselect` does. Set it to `always` if you want that behaviour.

**Comment by thallium** [comment]
> @thallium that's what `preselect` does. Set it to `always` if you want that behaviour.

Thank you. Didn't realize this : )

**Comment by wkcosmology** [comment]
Can we add a limit to the number of candidates? Sometimes there are dozens candidates and it undermine the speed.

**Comment by wkcosmology** [comment]
When I restrict the max_abbr_width to a small value, the signature showed by echodoc.vim plugin will also be truncated.
So when I want to show the complete signature, I need to make max_abbr_width to a large value.
![Screen Shot 2021-02-23 at 10 37 40 PM](https://user-images.githubusercontent.com/29885814/108944239-59112f80-7628-11eb-8927-18dabf8b843c.png)
![Screen Shot 2021-02-23 at 10 39 06 PM](https://user-images.githubusercontent.com/29885814/108944242-59a9c600-7628-11eb-9f59-8ba37ee50996.png)

**Comment by lourenci** [comment]
Show signature function while filling the parameters, like nvim-completion. I had to use nvim-completion in unknown codebases. I didn't find any plugin to do that. =/

**Comment by cloggier** [comment]
@lourenci @hrsh7th has already stated that this is out of scope for compe. You can use[ ray-x/lsp_signature.nvim](https://github.com/ray-x/lsp_signature.nvim) to display signature.

**Comment by cloggier** [comment]
@lourenci It doesn't work for me neither. It works on my machine with minimal config though. If you will still have a problem let's discuss it over there, maybe we can figure out why it doesn't work.

**Comment by thallium** [comment]
@lourenci  Just tried it, worked for me tho.

**Comment by lourenci** [comment]
@cloggier @thallium no way to get that plugin working. I've just tried... =/

It'd be so beautiful if @hrsh7th would implement here, like vim-lsp and nvim-completion. 😄

**Comment by LoricAndre** [comment]
I haven't seen it here, but emmet completion would be nice !

**Comment by cloggier** [comment]
@LoricAndre https://github.com/hrsh7th/nvim-compe/issues/190

**Comment by LoricAndre** [comment]
Thank you !

**Comment by goolord** [comment]
a source like the 'user' source using `completefunc` in completion-nvim https://github.com/nvim-lua/completion-nvim/blob/master/doc/completion-nvim.txt#L259

**Comment by CKolkey** [comment]
I think it would be very cool if, when inserting a suggestion, compe would take into account existing characters in the buffer. Here's what I mean by that:

Imagine you have the following text:
`my_cool_variable`

and you want to change it to match the previously used:
`my_awesome_variable`

As you remove `cool` and start typing `aw`, compe will (correctly) suggest `my_awesome_variable`.
The problem is, if you select this, the text in the buffer will be:
`my_awesome_variable_variable` because the existing `_variable` text was never removed, and not incorporated into the completion. 

![ezgif com-gif-maker](https://user-images.githubusercontent.com/7228095/112455695-4179b380-8d5a-11eb-8443-89f90b56d512.gif)

Compare this to how Deoplete handles the same situation:
![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/7228095/112456605-370be980-8d5b-11eb-8b3d-6e5df9167e7a.gif)

**Comment by tzachar** [comment]
> I think it would be very cool if, when inserting a suggestion, compe would take into account existing characters in the buffer. Here's what I mean by that:
> 
> Imagine you have the following text:
> `my_cool_variable`
> 
> and you want to change it to match the previously used:
> `my_awesome_variable`
> 
> As you remove `cool` and start typing `aw`, compe will (correctly) suggest `my_awesome_variable`.
> The problem is, if you select this, the text in the buffer will be:
> `my_awesome_variable_variable` because the existing `_variable` text was never removed, and not incorporated into the completion.
> 
> ![ezgif com-gif-maker](https://user-images.githubusercontent.com/7228095/112455695-4179b380-8d5a-11eb-8443-89f90b56d512.gif)
> 
> Compare this to how Deoplete handles the same situation:
> ![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/7228095/112456605-370be980-8d5b-11eb-8b3d-6e5df9167e7a.gif)

This is exatly what was requested in #157 
@hrsh7th is currently working on it.

**Comment by lbiaggi** [comment]
source representation customization, I mean something like [galaxy](https://github.com/glepnir/galaxyline.nvim) do with vim modes

**Comment by ii14** [comment]
@lbiaggi you can change `menu` option for each source. Is that what you mean? Changing `[LSP]`, `[Buffer]` etc. in completion menu?

**Comment by lbiaggi** [comment]
> @lbiaggi you can change `menu` option for each source. Is that what you mean? Changing `[LSP]`, `[Buffer]` etc. in completion menu?

Yup

**Comment by elianiva** [comment]
Try this

```lua
source = {
    buffer = {
        menu = "[FOO]"
    }
}
```

**Comment by lbiaggi** [comment]
Ok, I tried 
with 
```
let g:compe.source.buffer = v:true
let g:compe.source.buffer.menu = "[B]"
```
and did not worked.

I will try configuring using lua, thanks!

**Comment by hrsh7th** [comment]
```
let g:compe.source.buffer = {}
let g:compe.source.buffer.menu = "[B]"
```

**Comment by lbiaggi** [comment]
> ```
> let g:compe.source.buffer = {}
> let g:compe.source.buffer.menu = "[B]"
> ```

Ops, thanks! I though v:true was necessary... my bad

**Comment by lrvpxl** [comment]
Adding an option `max_items_visible` that restricts max number of items that could be visible in autosuggest but keeping the ability to navigate through the clipped/hidden items using `compe#scroll`.

**Comment by cloggier** [comment]
@lrvpxl use `pumheight`.

**Comment by PyGamer0** [comment]
How about more `Math functions` for calc completion.

Like:
- Trigonometry
- Constants like PI, E
- Nth Root
- Equations

**Comment by hrsh7th** [comment]
Lua has a `math` module. We can easily add `math.*` for calc completions.

**Comment by 6cdh** [comment]
Complete regular expression? It's a random idea. I don't know how to do it and what it should be.

**Comment by pedro757** [comment]
### Command line completion

Command line completion based on current buffer when searching words:

1. `:%s/somethingToComplete//g`
2. `/somethingToComplete`
3. `?somethingToComplete`

There are plugins like `vim-sherlock` and `gelguy/wilder.nvim` these try to do some of the features but they have a lot of bugs and they're not that good

**Comment by shaunsingh** [comment]
After using the plugin for a while my two requests would be: 

1. Delay before autocomplete option - it gets a bit distracting when its coming up as I type, maybe having a little delay (e.g. only show menu after 100ms of pause) would be nice. I imagine it would be lighter on the cpu as well
2. Command line completion - I'm terrible and remembering commands, it might be  a bit difficult but it would be awesome if possible

Lazy loading and better startup times would also be nice but the plugin is quite fast already, keep up the good work!

**Comment by matu3ba** [comment]
@shaunsingh Delay without explicit option looks like a broken language server. Also delay messes with the internal state of nvim-compe (what data should be kept and which dropped). Last, but not least, it is very much unclear at what timepoint nvim-compe should ask the lsp for answers.

Why does keypress for completion not work for you?

Take note that many lsps are often much slower than typing speed (~5ms per word or 1ms per character), because they work on AST data which are not cache-efficient due to necessary location lookups and missing lsp compiler infrastructure (location information need to be separately resolved).

> Command line completion

CLI completion requires shell CLI support, which in turn is poorly implemented with scripts for most (POSIX) shells.
If you want to make it properly consider porting [shellac](https://gitlab.redox-os.org/AdminXVII/shellac-server) to C. CLI arguments form a (relative simple) grammar on itself.

**Comment by shaunsingh** [comment]
Thank you for the quick reply. I think the correct term I meant to say was "command-mode completion, the one you enter after hitting the colon key. Sorry about the confusion.

**Comment by n3wborn** [comment]
I don't know if this is nvim-compe job but auto add "use ***" statements in php files is the only reason I still use Coc for the moment, I tried compe while moving to (almost) full lua config and really loved it but this really miss me :/
The day I leave Coc and fully embrace compe will be so good to me ^^

**Comment by elianiva** [comment]
compe can do that already -- the hardwork is handled by the language server like intelephense or phpactor and compe just displays the autocomplete

**Comment by n3wborn** [comment]
> compe can do that already -- the hardwork is handled by the language server like intelephense or phpactor and compe just displays the autocomplete

Thanks for the reply !
I don't know exactly what Intelephense does but use statements are only added when I use Coc, not with compe :/ . But if this is not a compe related problem (I assume Intelephense does the job but maybe I'm wrong...) I'll investigate a bit more. 
Thanks for all the amazing job :)

**Comment by fenuks** [comment]
Dictionary source would be nice (I'm aware of spell source, but it's something different, I mean completion one can invoke with <kbd>C-X C-K</kbd>). Deoplete [has one](https://github.com/deoplete-plugins/deoplete-dictionary), and there was some discussion on [reddit on the subject](https://libredd.it/r/neovim/comments/mrbyyu/nvimcompe_dictionary_completion/). I will try to write this source at the weekend.

**Comment by Gavinok** [comment]
> Dictionary source would be nice
Is something like this a good fit? https://github.com/Gavinok/compe-look

**Comment by fenuks** [comment]
Yes, thank you! I tried to write something on my own (not using look, but instead loading whole file into memory), but after few seconds it triggered error https://github.com/neovim/neovim/issues/11867, so it wasn't usable at all. Also reading file for the first time with `io` module blocked editor, even though I deferred it to `Source:complete` method.

Your source works perfect and is very smooth, albeit `look` doesn't return all results when prefix is only few letters long. It's not much of issue in practice, though.

**Comment by danielnehrig** [comment]
is it possible to add a border to the completion menu Pmenu

**Comment by laughingman-hass** [comment]
Not sure if it currently possible, as I couldn't find anything, but would really like to limit the number of candidates compe shows. Sometimes my floating menu covers the whole height of my screen which can be distracting.

**Comment by hrsh7th** [comment]
You can use `set pumheight=15`

**Comment by francisco285** [comment]
Some ideas

Sources:
- Abbreviations (:h Abbreviations)
- Digraphs (:h Digraphs)
- Quickfix/Location-list entries?
- Emoticons (similar to emojis, e.g. typing `;shr` should suggest `;shrug;` and expand to `¯\_(ツ)_/¯`. Of course ";" can be any other character)

Zsh-like completion (probably not useful at all):
![image](https://user-images.githubusercontent.com/77457027/124421341-a3718e00-dd37-11eb-994d-3bb5b73feec5.png)

**Comment by timsofteng** [comment]
> I think it would be very cool if, when inserting a suggestion, compe would take into account existing characters in the buffer. Here's what I mean by that:
> 
> Imagine you have the following text:
> `my_cool_variable`
> 
> and you want to change it to match the previously used:
> `my_awesome_variable`
> 
> As you remove `cool` and start typing `aw`, compe will (correctly) suggest `my_awesome_variable`.
> The problem is, if you select this, the text in the buffer will be:
> `my_awesome_variable_variable` because the existing `_variable` text was never removed, and not incorporated into the completion.
> 
> ![ezgif com-gif-maker](https://user-images.githubusercontent.com/7228095/112455695-4179b380-8d5a-11eb-8443-89f90b56d512.gif)
> 
> Compare this to how Deoplete handles the same situation:
> ![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/7228095/112456605-370be980-8d5b-11eb-8b3d-6e5df9167e7a.gif)

I would be great!

**Comment by n3wborn** [comment]
I know having a maximum number a candidates is already mentioned. What I think to be better is being able to have a **maximum candidates per sources**.

For example, 10 candidates for nvim-lsp, 5 for buffer, etc.

**Comment by hrsh7th** [comment]
@n3wborn It's easy but it doesn't improve performance because limitting items after `sort`.

**Comment by numToStr** [comment]
IDK if this is already implemented or not. But I would love to have completions for git contributors/authors name. This can be useful when assigning `TODO` or `FIXME`.

This could be included in the git source.

**Comment by ArenM** [comment]
I'd like to have the ability to insert all the text that's the same between all the completion options, a bit like how completion works in zsh, where the first time you press tab it completes everything that's the same and the next time it gives a list of options.

Even better would be if compe could have similar subsections of all the completion options as completion options, so for example if I have a set of functions that begin with `prog_namespace1` and another set that begin with `prog_namespace2` then after I type `prog_` the first 2 completion options would be `prog_namespace1` and `prog_namespace2` followed by all the functions from those 2 namespaces (assuming all the functions that begin with `prog_` are either in namespace 1 or  2).
